[RFC]ZG scripts by kraminpouds

Phatcat opened the pull request:
Some ZG script corrections by @kraminpouds

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cmangos/mangos-classic/144)

<!-- Reviewable:end -->

Konctantin commented on src/scriptdev2/scripts/eastern_kingdoms/zulgurub/boss_jeklik.cpp:
```diff
@@ -14,52 +14,52 @@
 
 enum
 {
-    SAY_AGGRO                   = -1309002,
```
Why have removed the alignment?

Phatcat replied:
No idea why'd he'd do so. I can write some fixup for these if we decide the code is good, though.

killerwife commented on src/scriptdev2/scripts/eastern_kingdoms/zulgurub/boss_jeklik.cpp:
```diff
@@ -183,6 +188,24 @@ struct boss_jeklikAI : public ScriptedAI
         }
     }
 
+    void DoSpellBombRiders()
```
Possibly review if targeting could be moved to core.

killerwife commented on src/scriptdev2/scripts/eastern_kingdoms/zulgurub/boss_jeklik.cpp:
```diff
@@ -215,24 +238,51 @@ struct boss_jeklikAI : public ScriptedAI
             if (m_uiSonicBurstTimer < uiDiff)
             {
                 if (DoCastSpellIfCan(m_creature, SPELL_SONIC_BURST) == CAST_OK)
+                {
```
this code style is crap :D

Phatcat replied:
Sadly it is, yeah.... so many unneeded diffs just to make the code style of the functions WORSE.... -.-

killerwife commented on src/scriptdev2/scripts/eastern_kingdoms/zulgurub/boss_jeklik.cpp:
```diff
@@ -215,24 +238,51 @@ struct boss_jeklikAI : public ScriptedAI
+                    {
+                        //DoCastSpellIfCan(pTarget, SPELL_SUMMON_FRENZIED_BATS);
+                        if (Creature* bat = m_creature->SummonCreature(NPC_FRENZIED_BAT, -12291.6220f, -1380.2640f, 144.8304f, 5.483f, TEMPSUMMON_DEAD_DESPAWN, 0)) {
+                            bat->SetLevitate(true);
```
not sure about classic, but in TBC SetLevitate doesnt exist

Phatcat replied:
The fact that it build with no issues should indicate that classic indeed has such a thing as SetLevitate. I would guess such a thing is not needed when you have flying in tbc and later, so maybe the tbc and wrath equivilant would be something like 'SetFlying' perhaps?

killerwife replied:
No, meaning was that SetLevitate in TBC and classic doesnt have a packet associated with it. I.E. IT IS supposed to do something clientside, but it doesnt update the way we are used to in WotLK. (or maybe im wrong and in classic/TBC its used only serverside)

xfurry replied:
Main problem here: why not use the spell, instead of the summon function?
We are using spells for a reason, and when we don't it's usually because it's really too complicated to fix the spell.

Phatcat replied:
@kraminpouds may not have been aware that there was a spell associated with the summon that is heppening here. This is not something he introduced: https://github.com/cmangos/mangos-classic/pull/144/files#diff-d8f78f7cf8c89e261352a50451fc09a0L291

I believe what you are saying but I don't like it. Nothing should be considered unfixable as far as I'm concerned, but do we know what spell this is regarding and if this spell even works or not?

xfurry commented:
Why are we doing this?
I don't understand what is the purpose this PR with very bad code style and a lot of hacks...

If we have issues with the ZG scripts, we should investigate them ourselves.
As far as I'm aware ZG scripts are pretty good.

killerwife commented:
@xfurry @Phatcat wanted me to provide him some feedback on these things, so i did, + I am not the boss on this one.

Phatcat commented:
@xfurry
I asked for feedback on this for this specific reason; I am not a scripter so I don't know what good scripting is really, in my eyes all scripting is hackfixing as it is all pretty case specific.

If you feel these were so bad, please always feel free to let me know about it, on the other hand, had some of this been useful, would the wiser decision of me to just have ignored it and thrown it away?

Phatcat commented:
and if we have no issues with ZG, why is this still open : https://github.com/cmangos/issues/issues/839 ?

xfurry commented:
Guys,

I don't like this kind of approach, when we see some code on github (which on top of everything, has chinese comments which we can't understand...) and then we make a PR out of it, just because it might help.

I would suggest the following steps:
1. If there is some issue identified on a script, first we need to do a deep investigation, to understand what is wrong and how can we fix.
2. If we find something on github that works as a fix for our problem, then we can make a PR with that. Otherwise we can try to fix it individually, or just report it on forum / github.



> and if we have no issues with ZG, why is this still open : cmangos/issues#839 ?

@Phatcat I don't know. I'm not an expert on ZG scripts :smile: 
But the fact that that ticket is still open, it doesn't mean that what is described in the ticket is actually right. Someone needs to do some proper research, and if we are good with the changes proposed by that ticket or this PR, then we can implement them, but in a proper and cleaner way.
Keep in mind that ZG script has changed in 4.1, so not everything which is reported recently is accurate.

xfurry commented:
So, don't take me wrong.
I like the fact that the community is active and willing to help.
But let's be somehow rational in this manner, and try to nail down the real problems first.

Phatcat commented:
I am being rational.
Just because these people are chinese doesn't mean they don't know what they're doing....

killerwife commented:
@Phatcat Well at least partially I have to agree with @xfurry , the styling in those commits is really atrocious :D (but for the rest im distancing myself because I hate boss scripting and im forced to do it on vengeance)

Phatcat commented:
@xfurry
If noone else cares to, then I'll review all of this imported code by myself. I was hoping I wasn't left with all of this to myself; there could be tons of reasons these people have not pr'ed even though their work is fine (now I am not talking this pr specifically, again, don't know scripting, but you not wanting to review them because the person who made them has not pr'ed them himself. Not enough to get me to pull them down, nope.

Phatcat commented:
@xfurry
It actually seems like we be having more than just one issue open for ZG: https://github.com/classicdb/database/issues?q=Zul%27+Gurub+is%3Aopen
Maybe it wouldn't be too bad looking this over.

cala commented:
I barely never looked at ZG yet (DB or core issues). I'm currently working on Stratholme and there is far more job than I expected. I'll look at ZG one day, but can't provide an ETA.

Also, most of these issues on Classic DB were opened by nekoi who made some valid reports but also some that were just based on wrong memories of the event. So checking the validity of those issues would be the first step, I think. Then, flag those related to DB, ACID, scripting or core. :)

xfurry commented:
Guys...
Instead of wasting time on this commit, which we have no clue about, why not have a look on the original SD2 code that we have, and fix whatever is wrong with our code directly?

I don't want to be disrespectful, but from my point of view there is nothing to review here.
This is just some custom code, with very bad style, that won't be accepted into the core mangos.

Here are my original commits on ZG, when we cleaned up the code a few years ago.
If there are any problems or bugs, let's investigate it on our code.

Jeklik: https://github.com/scriptdev2/scriptdev2/commit/8dd714a8fa649a3ec3acced439479541d33fa055
Venoxis: https://github.com/scriptdev2/scriptdev2/commit/bf53d7d0960eb28c7ebe218e9888aca4e4bcce7e
Marli: https://github.com/scriptdev2/scriptdev2/commit/3a6baa472c9c948c96926b85ba810d6ca921c65f
Arlokk: https://github.com/scriptdev2/scriptdev2/commit/17382dc9d32333b1c80cdb59df386767dff6e444

Phatcat commented:
@xfurry I don't think that's completely fair to @kraminpouds

If you look [here](https://github.com/scriptdev2/scriptdev2/commit/8dd714a8fa649a3ec3acced439479541d33fa055#diff-99875d8713837a2170d1c1f18701cd9dR38) and then look at this commit [here](https://github.com/cmangos/mangos-classic/pull/144/files#diff-d8f78f7cf8c89e261352a50451fc09a0R38), [here](https://github.com/cmangos/mangos-classic/pull/144/files#diff-d8f78f7cf8c89e261352a50451fc09a0R101) and [here](https://github.com/cmangos/mangos-classic/pull/144/files#diff-d8f78f7cf8c89e261352a50451fc09a0R263) it looks like kramin looked into and had a go at that. That doesn't really classify as custom in my book. Look [here](https://github.com/cmangos/mangos-classic/pull/144/files#diff-a9691ba789ccc41fcd51e9c8963b666eR238) as well. you think that is 'just some custom code' - this looks very proper to me, bad code style or not. It's not like code style is something the compiler cares about anyway, and I said already, I can fix the code style in no moment really.

Phatcat commented:
I honestly think it is you wasting your time trying to shut this PR down, because that is not going to happen. Untill you can deny the very fixes themselves, this ain't going down.

Phatcat commented:
It's also not that we couldn't find out if we really wanted; these were originally made as several different commits but I tried seperating the core issued them and squash to make it easier to review . . .
The commits can be found here https://github.com/kraminpouds/mangos-classic/commits/develop

https://github.com/kraminpouds/mangos-classic/commit/977ed1cd964f260a281709936e87a2c5851ad84e
https://github.com/kraminpouds/mangos-classic/commit/2a3d8c1db8f3f543338e876181ab63567ff1f117
https://github.com/kraminpouds/mangos-classic/commit/8dbd8343667c262773a7e4a1976c02720d138840
https://github.com/kraminpouds/mangos-classic/commit/6f602d5f40373d86c08c0c9f30d345ec72156896
https://github.com/kraminpouds/mangos-classic/commit/a5d4ee537f4e873edea4d2ce6da7992f4f2459e1

The core changes from these are in https://github.com/cmangos/mangos-classic/pull/139 ....

@keyshuwen could maybe tell us what all these commits say if we ask him nicely...

I really don't put in this effort for nothing, so don't deny this without at least taking an honest look at it.

Phatcat commented:
>https://github.com/kraminpouds/mangos-classic/commit/977ed1cd964f260a281709936e87a2c5851ad84e  -- (Optimization of progenitor Zul'Gurub Boss - High Priest Venoxis)
https://github.com/kraminpouds/mangos-classic/commit/2a3d8c1db8f3f543338e876181ab63567ff1f117  -- (Optimization of progenitor Zul'Gurub Boss - High Priestess Jeklik)
https://github.com/kraminpouds/mangos-classic/commit/8dbd8343667c262773a7e4a1976c02720d138840  -- (Repair boss crash progenitor Zul'Gurub)
https://github.com/kraminpouds/mangos-classic/commit/6f602d5f40373d86c08c0c9f30d345ec72156896 -- (Change Boss_Jeklik SpawnBatsTimer to shorter).
https://github.com/kraminpouds/mangos-classic/commit/a5d4ee537f4e873edea4d2ce6da7992f4f2459e1 -- (Continue to optimize the progenitor Zul'Gurub boss script.).

Thank you @keyshuwen :heart:

Phatcat commented on src/scriptdev2/scripts/eastern_kingdoms/zulgurub/boss_venoxis.cpp:
```diff
@@ -107,90 +144,97 @@ struct boss_venoxisAI : public ScriptedAI
             if (m_uiVenomSpitTimer < uiDiff)
             {
                 if (Unit* pTarget = m_creature->SelectAttackingTarget(ATTACKING_TARGET_RANDOM, 0))
-                {
```
Wow, would you look at that @xfurry, now this commit actually improves on the codestyle. you're welcome. Now would you please take this seriously?

Phatcat commented:
@cmangos Do we have any serious scripters here who ain't on a vendetta against me willing to look this over or am I going to have to study and research not only ZG but also scripting in order for me to review this myself?

Phatcat commented:
@Tobschinski Could I quite possibly get you to look this over? I know you've done a bit of scripting before, what's your honest opinion on this, are these changes all horribly wrong?

cyberium commented:
Well guys i also think this may at least need some feedback. Forget the code style for a moment.
@xfurry said he dont know enough ZG.

Phatcat commented:
https://github.com/cmangos/issues/issues/839

>SPELL_PSYHIC_SCREAM = 22884, // spell not confirmed - needs research (TROLL FORM ONLY)

>Source: http://www.wowhead.com/spell=22884/psychic-scream
>Confirmed!

>Mentioned in guide and notice it in the youtube video.
>Source:
>http://www.disciplesguild.com/forums/topic/end-game-instance-strategies-for-rogues/
>http://ru-mangos.ru/files/earlyServers/1.8.0/scripts/2006-04-19-Some%20Fixes%20for%20Boss%20And%20Instance%20Mob%20Spells.txt
>https://www.youtube.com/watch?v=8yI57bjb2Yw

Phatcat commented:
Looks like this guy actually did his research:
https://github.com/cmangos/issues/issues/839

I think Kramin took advantage of that and actually tried to work in his feedback, no?

cala commented:
Hi,

I agree that this PR will probably never be merged as it is because of the code style and, like the PR for Sunken Temple, it rewrites parts of the existing code to match a different development approach than the one we are using here (like 100% focusing on a Classic approach while we are developing also for TBC, WotLK and Cata).

However, there are issues in ZG (valid or not, I can't tell for now) and maybe some of them are fixed/fixable in the current PR.
Moreover, there seems indeed to be there a good bunch of research made, thanks @Phatcat for compiling them.
So it would be a good idea to use that PR (or at least part of it) as a possible pool of ideas for fixing the current opened issues.

As I said before:
> I'm currently working on Stratholme and there is far more job than I expected.

I spent many **days** just doing my researches for Stratholme and I want to finish my work with that first to avoid doing the tedious research/reading again.

Once done, I can give a quick look at this PR to see if something matches one of the currently open PR. No ETA though and that won't be a whole rework of the dungeon either: I have already many things in my ToDo list to take care of before a (yet schedule) full revamp of ZG.

Phatcat commented:
@cala I would just like to tell you that I took the time to clean up the codestyle. Several days ago. Furthermore, if you guys keep repeating the codestyle as an excuse as to anything really you are kind of outright ignoring me and the effort I put in. Please don't do that.

> it rewrites parts of the existing code to match a different development approach than the one we are using here

No, it rewrites parts of the script that is broken in an attempt to fix it. Because I'm looking at it and it looks pretty proper, non-hackfixing and in no way out of style with the rest of the scripting done in the rest of the files. What exactly do you mean 'to match a different development approach'? Are we talking about my way of working here? Because that has nothing to do with this code.

cmangos/issues#839 - I'm willing to bet this guy spend days on that research as well, if not then at least **hours**. And afterwards Kramin spend hours on fixing these things according to his feedback.

cala commented:
> What exactly do you mean 'to match a different development approach'? Are we talking about my way of working here? Because that has nothing to do with this code.

@Phatcat Please, you should really start to consider criticism about other peoples' work/code as something that can be constructive in order to achieve a even better code/way to do things and not necessary as "this is bad, so you are/he is bad". Moreover, after reading again my previous comment, I still can't spot the place where I could have linked what is in those commits with your work. In fact, I always tried to trust you and push you forward so I find this kind of comments are disheartening.

By a different approach, I was referring to the fact that from what I read *quickly* in the PR (the stress here on "quickly" is important), it feels like @kraminpouds increases the use of spells ID and even hardcoded ones in some places. This can be perfectly fine for Classic but could make the code harder to port to TBC and later in case where the spells were removed or changed for those expansions. This happened to many scripts before, so we need to anticipate such issues, especially when the backporting guy is not always the script's author. That's what I was referring to: single expansion approach versus multi-expansions.

> cmangos/issues#839 - I'm willing to bet this guy spend days on that research as well, if not then at least hours. And afterwards Kramin spend hours on fixing these things according to his feedback.

I do not deny nor undervalue the time it took to do this and I never did. I know too well the time it takes to research specific things about a game released 12 years ago and that had dozen of patches since.
I was just stating that I don't see that as a reason to throw away the time I invested myself in my current work to instead spend hours reading, understanding and checking the ZG research "now". So instead of "now", it will be "later" because this is how I work: I prefer to focus on a few things and having them well done in a reasonable amount of time, rather than scatter myself on several at once and taking ages to only deliver a few. Some people can do it, I can't.

So, again: yes, there are interesting things in this PR (I did not read it enough to be sure, but I trust you on that). However, I don't have the time to look at this right now, yet I will do it later if no one does it before.

Phatcat commented:
But you are not really being constructive are you? you say bad code style without saying what needs to be fixed about it, so when I try to fix the code style I'd think people would actually notice, else I am starting to find that argument a bit hollow honestly, and it is the only real argument here as far as I can tell?

the reason you can't see my commits are because I do fixups because this is not my work so I don't want to take credit and neither do I wantto bloat our repo...... you think I put zero work in this and just pulled them straight in?

Here they are then:

kraminpouds@977ed1c 尝试优化祖尔格拉布boss——毒蛇
kraminpouds@2a3d8c1 尝试优化祖尔格拉布boss——蝙蝠
kraminpouds@8dbd834 尝试修复祖尔格拉布boss crash
kraminpouds@6f602d5 Change Boss_Jeklik SpawnBatsTimer to shorter (:+1:)
kraminpouds@a5d4ee5 继续优化祖尔格拉布boss脚本 

Lie to everyone here and say there is no difference in codestyle or content between these commits and the one I have PR'd here.

I'm not having a ago at anyone but noone here except me and cyberium, and you ever so slightly, seems to see the potential in what Kraminpouds have done here.

Phatcat commented:
Neither did I ever state Kraminpouds work was perfect...... This is why I've been asking for reviewing.... I'm starting to feel a bit hopeless here.

Phatcat commented:
@cala here is how I saw the link between this and my work by the way " different development approach " - this is not a development approach, this is a scripting approach, what I am doing right here, is a development approach however. If you only meant this commit then I misunderstood you. I'm sorry.

cala commented:
> you say bad code style without saying what needs to be fixed about it

Because saying point by point what is to be fixed would be reviewing the PR and, again, I don't have the time to do it currently so I only give generic directions/indications hoping that at least would be more helpful or encouraging than saying nothing and letting people believe we don't care about their contributions. I guess I was wrong...

> I am starting to find that argument [code style] a bit hollow honestly, and it is the only real argument here as far as I can tell?

As you seem to need a precise example: https://github.com/cmangos/mangos-classic/pull/144/files#diff-d8f78f7cf8c89e261352a50451fc09a0R155
```C++
pTarget->CastSpell(pTarget, 23971, true);
```
This line of code may work perfectly for Classic as this code is intended for that core but hardcoded values should be avoided because:
- it is hard to tell what spell is casted and what it does with only the spell ID instead of using a `enum` like we usually do. This makes the real meaning of that line of code hard to understand
- it makes the code harder to debug : let's imagine there was a typo in typing 23971 instead of 23917, yet spell 23971 is successfully casted but has no effect/visual. You won't get an error and the bug will be hard to spot because that line of code worked as expected. Chances of this are lesser to occur with enum
- if I backport it to TBC and that spell was modified/deleted, I will have to search for every occurrence of it in the code and change it instead of just having to update the enum

> But you are not really being constructive are you?

I don't know how I can be more constructive than what I said before: I will look at it.

Phatcat commented:
But that has as far as I know... Nothing to do with codestyle...
this is what I always considered codestyle; https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md - I come from Dolphin, the terms don't always mean the same everywhere..

While what you are talking about is what I consider a method, and I never claimed any method here was solid or good.... Cause I'm no scripter...

Phatcat commented:
I just fixed any 'hard-casting' being done (there was actually only 2 occurences of the same spell, which I added to the enums)

Codestyle is now slightly improved as well. I looked it all over and couldn't find anything else that seemed out of place with the rest of the scripting being done here.

Please don't hate me @cala , it's not like I want to give you more work I'm trying to be as helpful as possible... we just got another issue report for ZG: http://cmangos.net/thread-7488.html, this time with a crash that Kramin ´could have possibly already fixed: https://github.com/cmangos/mangos-classic/pull/139/commits/e0f157cdb0958039c33ba1dbc2b417f30cebe8a6

Phatcat commented:
@VladimirMangos Just look here and know that I'm not trying to sweep ZG (or any other) issues under the rug, I just think people should be able to 'enjoy' a broken ZG without crashing while we work things out behind the scenes.

I don't know if you're big on scripting but if you wouldn't mind I would love your view on these changes, really.

VladimirMangos commented:
Sorry, know very limited in script part and DB development.

Phatcat commented:
Oh, that's okay, I know no scripting whatsoever.

I think it may be time I accept that this is going stale until a scripter is willing to pick up ZG. No hard feelings here, I just think it's a bit of a shame, but the real shame may be in the lack of scripters on our team.

Tobschinski commented:
I don't think I am useful here, I don't have any sniffs from ZG.

kraminpouds commented:
@Phatcat @xfurry 
I am sorry for my bad code style and a lot of hacks. I did this just for fun in spare time, when I play with my friends. I never thought you would focus on. I am not a really coder, also don't have a clear understanding of the core run. so...

Phatcat commented:
> I am not a really coder

Neither am I, don't tell anyone ;)

no, seriously, don't apologize for any of that; as you said, you did this for fun, for you and some friends, so what code style could anyone expect you to live up to, really.

You don't need a clear understanding to put in good honest work; noone here has a 'clear' understanding of the core; we all understand some parts better than others, and that's pretty much it.

I don't know why you say lots of hacks, I don't really see much hacking in what you've done; actually it looks like quite the opposite, that you actually look your time to do things properly.

But now that we have you here; let me take the opportunity to ask you a few questions about this one:

- Why did you change the things you did?
- Did you change anything with the intention of making something 'custom' or less 'blizzlike'?

kraminpouds commented:
> Why did you change the things you did?
Just my friends said it don't run like their memory.

> Did you change anything with the intention of making something 'custom' or less 'blizzlike'?
I tried to change it 'blizzlike' so that my friends and I can have a sense, but something can not to confirm. so I can only adjust by the memory of my friends or me and some videos.

So, all changes are only for fun with my friends.  ^ ^;

cyberium commented:
we can close this as for sure it will not be merged that way. At least if someone want to fix ZG when time is come it can use all of this data
